spark2/src/main/java/org/apache/iceberg/spark/source/Reader.java

```diff
@@ -132,7 +132,7 @@
     this.splitLookback = options.get(SparkReadOptions.LOOKBACK).map(Integer::parseInt).orElse(null);
     this.splitOpenFileCost = options.get(SparkReadOptions.FILE_OPEN_COST).map(Long::parseLong).orElse(null);
 
-    if (table.io() instanceof HadoopFileIO) {
+    if (table.io() instanceof HadoopConfigurable) {
```

Comment:
I don't think this is correct. The logic in this `if` statement assumes that it can create a `FileSystem` for the table's location if its `FileIO` is a `HadoopFileIO`. That's not necessarily the case if the io is just `HadoopConfigurable` because I might have my own implementation that for some reason uses a Hadoop conf.

Reply:
This seems to be related to the discussion in the dev list regarding the locality read configuration. In the use case I am trying to support, it actually needs to run that code path and turn locality read to true by default although it is not a `HadoopFileIO`. So we have both cases to support, and it is not sufficient to determine preference of locality read purely based on the FileIO implementation and file URI. 

In the code path, because it is at reader initialization time, table property seems to be the best place to store this default behavior, although I agree this is not really elegant as it is spark specific.

For now, I think using the `HadoopConfigurable` check instead of `HadoopFileIO` check is more flexible, because users who do not need locality for `HadoopConfigurable` are likely not using HDFS anyway and will fail the check, and they can also use `locality` option to override the choice to turn it off.

Reply:
Actually, instead of table property, Hadoop configuration should be a better place to put this default. Because it requires a file system, a Hadoop configuration is needed anyway in the code path. This can avoid placing Spark specific configs in table property.

Reply:
I think it's a reasonable idea to be able to set a property somewhere and use locality. But this specifically should check for `HadoopFileIO` until we know what that property does and where it is because the locality code path is going to get a `FileSystem` instance. If you aren't using `HadoopFileIO` then there is no guarantee that we can get a file system or that it is the same as the IO instance.

Let's fix the locality problem later and keep this commit focused on updates to add `HadoopConfigurable`. With just that as the goal of this PR, I think this should be left as `HadoopFileIO`.